frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -0,0 +1,103 @@
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.lang.RuntimeException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <tripId, placeId, PlaceVisitModel>
+  Map<String, Map<String, PlaceVisitModel>> storage = new ConcurrentHashMap<>();
+
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit)
+      throws PlaceVisitAlreadyExistsException, Throwable {
+    try {
+      storage.compute(placeVisit.tripId(), (tripKey, placesMap) -> {
+        if (placesMap == null) {
+          Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+          newPlaceMap.put(placeVisit.placeId(), placeVisit);
+          return newPlaceMap;
+        }
+        if (placesMap.get(placeVisit.placeId()) != null) {
+          throw new RuntimeException(new PlaceVisitAlreadyExistsException("PlaceVisit "
+              + placeVisit.name() + " already exists for trip " + placeVisit.tripId()));
+        }
+        placesMap.put(placeVisit.placeId(), placeVisit);
+        return placesMap;
+      });
+    } catch (RuntimeException e) {
+      throw e.getCause();
+    }
+  }
+
+  @Override
+  public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+
+    if (placesMap.remove(placeId) == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+  }
+
+  @Override
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+    return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
+  }
+
+  @Override
+  public boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      return false;
+    }
+
+    PlaceVisitModel originalPlace = placesMap.computeIfPresent(
+        placeId, (placeKey, place) -> place.toBuilder().setUserMark(newStatus).build());
+    if (originalPlace == null) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public List<PlaceVisitModel> getTripPlaceVisits(String tripId) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    List<PlaceVisitModel> tripPlaceVisits = new ArrayList<>();
+    if (placesMap == null) {
+      return tripPlaceVisits;
+    }
+    for (PlaceVisitModel place : placesMap.values()) {
+      if (place != null
+          && (place.userMark().equals("must-see") || place.userMark().equals("if-time"))) {
```

Comment:
You should probably document these predicates in the interface method description since the behavior is not otherwise obvious to callers.